report/04_implementation/01_flutter_application.md

```diff
@@ -6,23 +6,149 @@ web application) with a single code base.
+class SoTiredApp extends StatelessWidget {
+  const SoTiredApp({Key? key}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    
+    ...
+    
+    // Register the Serviceprovider via the ChangeNotifierProvider provided 
+    // by the provider package.
+    return ChangeNotifierProvider(
+      create: (BuildContext context) => ServiceProvider(),
+      child: const SoTiredAppContent(),
+    );
+  }
+}
+
+...
+
+class SoTiredAppState extends State<SoTiredAppContent>
+    with WidgetsBindingObserver {
+  
+  ...
+
+  // This method is called immediately after every app start
+  @override
+  void initState() {
+    // call super method and instantiate widget observer
+
+    // Provider.of is part of the provider package and returns a registered
+    // provider. After that you can call every method this provider contains.
+    // The unnamed method the init function receives will set the variable 
+    // _doneInitializing to true which allows the typical Flutter build method 
+    // to ensure that every key has been initialized and returns the actual 
+    // app.
+    Provider.of<ServiceProvider>(context, listen: false)
+        .init(() => setState(() => _doneInitializing = true), path);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    if (_doneInitializing) {
+      return MaterialApp(...);
+    }
+  }
+
+  ...
+
+}
+
+...
+
+class ServiceProvider extends ChangeNotifier {
+  final ConfigManager _configManager = ConfigManager();
+  final DatabaseManager _databaseManager = DatabaseManager();
+  final Notifications _notifications = Notifications();
+
+  // ...
+  // Getter and Setter
+  // ...
+
+  Future<void> init(Function onDoneInitializing, <more arguments> ...) async {
+    // ...
+    // initialization tasks
+    // ...
+
+    onDoneInitializing(); // <- invoke this method after all init tasks have 
+                          // completed
+  }
+}
+```
+
+### Configuration Manager
+
+As mentioned in [Service Provider](service-provider) the configuration manager 
+is part of it. It also implements the Singleton design pattern principle to 
+ensure only one instance of itself and only one instance of client config 
+exists. The config manager provides various functions to write and read data 
+to / from a local json file, although its main purpose is fetching a 
+configuration from the study server. If no server can be reached and / or no 
+local json file exists, it holds a default configuration which can be loaded 
+any time.
+
+### Database Manager
+
+The database package is a relatively big package, as it holds multiple models 
+and their generated adapters. As every key element of the app it is also part 
+of the service provider class.  
+As most databases it contains basic CRUD (Create, Read, Update, Delete) 
+operations and also provides the ability to export all database contents as 
+json, so it can be send to the server. Mentionable when using *Hive* as 
```

Comment:
```suggestion
json, so they can be send to the server. Mentionable when using *Hive* as 
```

Reply:
Imo `it` is a better fit, because it is referring to one json object / file containing all database entries 🤔